File: packages/alpaca-option/src/numeric.ts
import { fail } from './error';

const SQRT_2 = Math.sqrt(2);
const SQRT_2PI = Math.sqrt(2 * Math.PI);
const SQRT_PI = Math.sqrt(Math.PI);
const ERF_EPSILON = 1e-18;
const ERF_MAX_ITERATIONS = 200;

function ensureFinite(name: string, value: number): void {
  if (!Number.isFinite(value)) {
    fail('invalid_numeric_input', `${name} must be finite: ${value}`);
  }
}

export type RangeExtrema = {
  minSpot: number;
  minValue: number;
  maxSpot: number;
  maxValue: number;
};

function erfSeries(x: number): number {
  const sign = x < 0 ? -1 : 1;
  const absX = Math.abs(x);
  let term = absX;
  let total = absX;

  for (let n = 1; n < ERF_MAX_ITERATIONS; n += 1) {
    term *= -(absX * absX) / n;
    const delta = term / (2 * n + 1);
    total += delta;
    if (Math.abs(delta) < ERF_EPSILON) {
      break;
    }
  }

  return sign * 2 * total / SQRT_PI;
}

function normalCdfTail(x: number): number {
  let fraction = 0;
  for (let k = ERF_MAX_ITERATIONS; k >= 1; k -= 1) {
    fraction = k / (x + fraction);
  }
  return normalPdf(x) / (x + fraction);
}

export function normalCdf(x: number): number {
  if (x === 0) {
    return 0.5;
  }

  if (Math.abs(x) <= 4) {
    return 0.5 * (1 + erfSeries(x / SQRT_2));
  }

  const tail = normalCdfTail(Math.abs(x));
  return x > 0 ? 1 - tail : tail;
}

export function normalPdf(x: number): number {
  return Math.exp(-0.5 * x * x) / SQRT_2PI;
}

export function round(value: number, decimals: number): number {
  ensureFinite('value', value);
  if (!Number.isInteger(decimals) || decimals < 0) {
    fail('invalid_numeric_input', `decimals must be a non-negative integer: ${decimals}`);
  }
  const multiplier = 10 ** decimals;
  return Math.round(value * multiplier) / multiplier;
}

export function linspace(start: number, end: number, count: number): number[] {
  ensureFinite('start', start);
  ensureFinite('end', end);
  if (!Number.isInteger(count) || count <= 0) {
    fail('invalid_numeric_input', `count must be a positive integer: ${count}`);
  }
  if (count === 1) {
    return [start];
  }

  const step = (end - start) / (count - 1);
  return Array.from({ length: count }, (_, index) => start + step * index);
}

export function brentSolve(
  lowerBound: number,
  upperBound: number,
  evaluate: (x: number) => number,
  toleranceInput?: number,
  maxIterationsInput?: number,
): number {
  ensureFinite('lowerBound', lowerBound);
  ensureFinite('upperBound', upperBound);
  if (lowerBound >= upperBound) {
    fail('invalid_numeric_input', `lowerBound must be less than upperBound: ${lowerBound} >= ${upperBound}`);
  }

  const tolerance = toleranceInput ?? 1e-10;
  if (!Number.isFinite(tolerance) || tolerance <= 0) {
    fail('invalid_numeric_input', `tolerance must be positive: ${tolerance}`);
  }

  const maxIterations = maxIterationsInput ?? 100;
  if (!Number.isInteger(maxIterations) || maxIterations <= 0) {
    fail('invalid_numeric_input', `maxIterations must be a positive integer: ${maxIterations}`);
  }

  let a = lowerBound;
  let b = upperBound;
  let fa = evaluate(a);
  let fb = evaluate(b);
  ensureFinite('f(lowerBound)', fa);
  ensureFinite('f(upperBound)', fb);

  if (Math.abs(fa) <= tolerance) {
    return a;
  }
  if (Math.abs(fb) <= tolerance) {
    return b;
  }
  if (fa * fb > 0) {
    fail('root_not_bracketed', `root is not bracketed: f(${a})=${fa}, f(${b})=${fb}`);
  }

  if (Math.abs(fa) < Math.abs(fb)) {
    [a, b] = [b, a];
    [fa, fb] = [fb, fa];
  }

  let c = a;
  let fc = fa;
  let d = b - a;
  let mflag = true;

  for (let iteration = 0; iteration < maxIterations; iteration += 1) {
    let s = (fa !== fc && fb !== fc)
      ? (a * fb * fc) / ((fa - fb) * (fa - fc))
        + (b * fa * fc) / ((fb - fa) * (fb - fc))
        + (c * fa * fb) / ((fc - fa) * (fc - fb))
      : b - (fb * (b - a)) / (fb - fa);

    const lowerWindow = (3 * a + b) / 4;
    const outsideWindow = a < b
      ? s <= lowerWindow || s >= b
      : s >= lowerWindow || s <= b;
    const cond2 = mflag && Math.abs(s - b) >= Math.abs(b - c) / 2;
    const cond3 = !mflag && Math.abs(s - b) >= Math.abs(c - d) / 2;
    const cond4 = mflag && Math.abs(b - c) < tolerance;
    const cond5 = !mflag && Math.abs(c - d) < tolerance;

    if (outsideWindow || cond2 || cond3 || cond4 || cond5) {
      s = (a + b) / 2;
      mflag = true;
    } else {
      mflag = false;
    }

    const fs = evaluate(s);
    ensureFinite('f(candidate)', fs);

    d = c;
    c = b;
    fc = fb;

    if (fa * fs < 0) {
      b = s;
      fb = fs;
    } else {
      a = s;
      fa = fs;
    }

    if (Math.abs(fa) < Math.abs(fb)) {
      [a, b] = [b, a];
      [fa, fb] = [fb, fa];
    }

    if (Math.abs(fb) <= tolerance || Math.abs(b - a) <= tolerance) {
      return b;
    }
  }

  fail('root_not_converged', `root solver did not converge in ${maxIterations} iterations`);
}

export function refineBracketedRoot(
  lowerBound: number,
  upperBound: number,
  evaluate: (x: number) => number,
  toleranceInput?: number,
  maxIterationsInput?: number,
): number {
  return brentSolve(lowerBound, upperBound, evaluate, toleranceInput, maxIterationsInput);
}

export function evaluatePoints(points: number[], evaluate: (x: number) => number): number[] {
  return points.map((point) => {
    ensureFinite('point', point);
    const value = evaluate(point);
    ensureFinite('f(point)', value);
    return value;
  });
}

export function scanRangeExtrema(
  lowerBound: number,
  upperBound: number,
  stepInput: number | undefined,
  evaluate: (x: number) => number,
): RangeExtrema {
  ensureFinite('lowerBound', lowerBound);
  ensureFinite('upperBound', upperBound);
  if (lowerBound > upperBound) {
    fail('invalid_numeric_input', `lowerBound must be less than or equal to upperBound: ${lowerBound} > ${upperBound}`);
  }

  const step = stepInput ?? 1;
  if (!Number.isFinite(step) || step <= 0) {
    fail('invalid_numeric_input', `step must be positive: ${step}`);
  }

  let spot = lowerBound;
  let value = evaluate(spot);
  ensureFinite('f(lowerBound)', value);

  const extrema: RangeExtrema = {
    minSpot: spot,
    minValue: value,
    maxSpot: spot,
    maxValue: value,
  };

  while (spot < upperBound) {
    spot = Math.min(spot + step, upperBound);
    value = evaluate(spot);
    ensureFinite('f(point)', value);
    if (value < extrema.minValue) {
      extrema.minValue = value;
      extrema.minSpot = spot;
    }
    if (value > extrema.maxValue) {
      extrema.maxValue = value;
      extrema.maxSpot = spot;
    }
  }

  return extrema;
}

File: crates/alpaca-option/src/numeric.rs
use crate::error::{OptionError, OptionResult};
use alpaca_core::float;

const SQRT_2: f64 = 1.414_213_562_373_095_1;
const SQRT_2PI: f64 = 2.506_628_274_631_000_2;
const SQRT_PI: f64 = 1.772_453_850_905_516;
const ERF_EPSILON: f64 = 1e-18;
const ERF_MAX_ITERATIONS: usize = 200;

fn ensure_finite(name: &str, value: f64) -> OptionResult<()> {
    if value.is_finite() {
        Ok(())
    } else {
        Err(OptionError::new(
            "invalid_numeric_input",
            format!("{name} must be finite: {value}"),
        ))
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct RangeExtrema {
    pub min_spot: f64,
    pub min_value: f64,
    pub max_spot: f64,
    pub max_value: f64,
}

fn erf_series(x: f64) -> f64 {
    let sign = if x < 0.0 { -1.0 } else { 1.0 };
    let abs_x = x.abs();
    let mut term = abs_x;
    let mut total = abs_x;

    for n in 1..ERF_MAX_ITERATIONS {
        term *= -(abs_x * abs_x) / n as f64;
        let delta = term / (2 * n + 1) as f64;
        total += delta;
        if delta.abs() < ERF_EPSILON {
            break;
        }
    }

    sign * 2.0 * total / SQRT_PI
}

fn normal_cdf_tail(x: f64) -> f64 {
    let mut fraction = 0.0;
    for k in (1..=ERF_MAX_ITERATIONS).rev() {
        fraction = k as f64 / (x + fraction);
    }
    normal_pdf(x) / (x + fraction)
}

pub fn normal_cdf(x: f64) -> f64 {
    if x == 0.0 {
        return 0.5;
    }

    if x.abs() <= 4.0 {
        return 0.5 * (1.0 + erf_series(x / SQRT_2));
    }

    let tail = normal_cdf_tail(x.abs());
    if x > 0.0 { 1.0 - tail } else { tail }
}

pub fn normal_pdf(x: f64) -> f64 {
    (-0.5 * x * x).exp() / SQRT_2PI
}

pub fn round(value: f64, decimals: u32) -> OptionResult<f64> {
    ensure_finite("value", value)?;
    Ok(float::round(value, decimals))
}

pub fn linspace(start: f64, end: f64, count: usize) -> OptionResult<Vec<f64>> {
    ensure_finite("start", start)?;
    ensure_finite("end", end)?;
    if count == 0 {
        return Err(OptionError::new(
            "invalid_numeric_input",
            "count must be greater than zero",
        ));
    }
    if count == 1 {
        return Ok(vec![start]);
    }

    let step = (end - start) / (count - 1) as f64;
    Ok((0..count).map(|index| start + step * index as f64).collect())
}

fn validate_brent_params(
    lower_bound: f64,
    upper_bound: f64,
    tolerance: Option<f64>,
    max_iterations: Option<usize>,
) -> OptionResult<(f64, usize)> {
    ensure_finite("lower_bound", lower_bound)?;
    ensure_finite("upper_bound", upper_bound)?;
    if lower_bound >= upper_bound {
        return Err(OptionError::new(
            "invalid_numeric_input",
            format!("lower_bound must be less than upper_bound: {lower_bound} >= {upper_bound}"),
        ));
    }

    let tolerance = tolerance.unwrap_or(1e-10);
    if !tolerance.is_finite() || tolerance <= 0.0 {
        return Err(OptionError::new(
            "invalid_numeric_input",
            format!("tolerance must be positive: {tolerance}"),
        ));
    }

    let max_iterations = max_iterations.unwrap_or(100);
    if max_iterations == 0 {
        return Err(OptionError::new(
            "invalid_numeric_input",
            "max_iterations must be greater than zero",
        ));
    }

    Ok((tolerance, max_iterations))
}

fn brent_solve_impl<F>(
    lower_bound: f64,
    upper_bound: f64,
    evaluate: F,
    tolerance: Option<f64>,
    max_iterations: Option<usize>,
) -> OptionResult<f64>
where
    F: Fn(f64) -> OptionResult<f64>,
{
    let (tolerance, max_iterations) =
        validate_brent_params(lower_bound, upper_bound, tolerance, max_iterations)?;

    let mut a = lower_bound;
    let mut b = upper_bound;
    let mut fa = evaluate(a)?;
    let mut fb = evaluate(b)?;
    ensure_finite("f(lower_bound)", fa)?;
    ensure_finite("f(upper_bound)", fb)?;

    if fa.abs() <= tolerance {
        return Ok(a);
    }
    if fb.abs() <= tolerance {
        return Ok(b);
    }
    if fa * fb > 0.0 {
        return Err(OptionError::new(
            "root_not_bracketed",
            format!("root is not bracketed: f({a})={fa}, f({b})={fb}"),
        ));
    }

    if fa.abs() < fb.abs() {
        std::mem::swap(&mut a, &mut b);
        std::mem::swap(&mut fa, &mut fb);
    }

    let mut c = a;
    let mut fc = fa;
    let mut d = b - a;
    let mut mflag = true;

    for _ in 0..max_iterations {
        let mut s = if fa != fc && fb != fc {
            a * fb * fc / ((fa - fb) * (fa - fc))
                + b * fa * fc / ((fb - fa) * (fb - fc))
                + c * fa * fb / ((fc - fa) * (fc - fb))
        } else {
            b - fb * (b - a) / (fb - fa)
        };

        let lower_window = (3.0 * a + b) / 4.0;
        let outside_window = if a < b {
            s <= lower_window || s >= b
        } else {
            s >= lower_window || s <= b
        };
        let cond2 = mflag && (s - b).abs() >= (b - c).abs() / 2.0;
        let cond3 = !mflag && (s - b).abs() >= (c - d).abs() / 2.0;
        let cond4 = mflag && (b - c).abs() < tolerance;
        let cond5 = !mflag && (c - d).abs() < tolerance;

        if outside_window || cond2 || cond3 || cond4 || cond5 {
            s = (a + b) / 2.0;
            mflag = true;
        } else {
            mflag = false;
        }

        let fs = evaluate(s)?;
        ensure_finite("f(candidate)", fs)?;

        d = c;
        c = b;
        fc = fb;

        if fa * fs < 0.0 {
            b = s;
            fb = fs;
        } else {
            a = s;
            fa = fs;
        }

        if fa.abs() < fb.abs() {
            std::mem::swap(&mut a, &mut b);
            std::mem::swap(&mut fa, &mut fb);
        }

        if fb.abs() <= tolerance || (b - a).abs() <= tolerance {
            return Ok(b);
        }
    }

    Err(OptionError::new(
        "root_not_converged",
        format!("root solver did not converge in {max_iterations} iterations"),
    ))
}

pub fn brent_solve<F>(
    lower_bound: f64,
    upper_bound: f64,
    evaluate: F,
    tolerance: Option<f64>,
    max_iterations: Option<usize>,
) -> OptionResult<f64>
where
    F: Fn(f64) -> f64,
{
    brent_solve_impl(
        lower_bound,
        upper_bound,
        |spot| Ok(evaluate(spot)),
        tolerance,
        max_iterations,
    )
}

pub fn refine_bracketed_root<F>(
    lower_bound: f64,
    upper_bound: f64,
    evaluate: F,
    tolerance: Option<f64>,
    max_iterations: Option<usize>,
) -> OptionResult<f64>
where
    F: Fn(f64) -> OptionResult<f64>,
{
    brent_solve_impl(
        lower_bound,
        upper_bound,
        evaluate,
        tolerance,
        max_iterations,
    )
}

pub fn evaluate_points<F>(points: &[f64], evaluate: F) -> OptionResult<Vec<f64>>
where
    F: Fn(f64) -> OptionResult<f64>,
{
    let mut values = Vec::with_capacity(points.len());
    for point in points {
        ensure_finite("point", *point)?;
        let value = evaluate(*point)?;
        ensure_finite("f(point)", value)?;
        values.push(value);
    }
    Ok(values)
}

pub fn scan_range_extrema<F>(
    lower_bound: f64,
    upper_bound: f64,
    step: Option<f64>,
    evaluate: F,
) -> OptionResult<RangeExtrema>
where
    F: Fn(f64) -> OptionResult<f64>,
{
    ensure_finite("lower_bound", lower_bound)?;
    ensure_finite("upper_bound", upper_bound)?;
    if lower_bound > upper_bound {
        return Err(OptionError::new(
            "invalid_numeric_input",
            format!("lower_bound must be less than or equal to upper_bound: {lower_bound} > {upper_bound}"),
        ));
    }

    let step = step.unwrap_or(1.0);
    if !step.is_finite() || step <= 0.0 {
        return Err(OptionError::new(
            "invalid_numeric_input",
            format!("step must be positive: {step}"),
        ));
    }

    let mut spot = lower_bound;
    let mut value = evaluate(spot)?;
    ensure_finite("f(lower_bound)", value)?;

    let mut extrema = RangeExtrema {
        min_spot: spot,
        min_value: value,
        max_spot: spot,
        max_value: value,
    };

    while spot < upper_bound {
        spot = (spot + step).min(upper_bound);
        value = evaluate(spot)?;
        ensure_finite("f(point)", value)?;
        if value < extrema.min_value {
            extrema.min_value = value;
            extrema.min_spot = spot;
        }
        if value > extrema.max_value {
            extrema.max_value = value;
            extrema.max_spot = spot;
        }
    }

    Ok(extrema)
}
